File: src/lexer.cpp
#include <cmath>
#include <string>
#include <vector>

#include "lexer.h"

// Current char
static char getChar(stream_t &stream) {
    if (stream.str.size() <= stream.at) {
        return EOF;
    }
    return stream.str[stream.at];
}

// Advance char
static char getNextChar(stream_t &stream) {
    if (getChar(stream) == '\n') {
        stream.lastLinePos = stream.at + 1;
        stream.line++;
    }
    stream.at++;
    char ret = getChar(stream);
    return ret;
}

// Advance only if char matches
static bool isNext(stream_t &stream, char c) {
    if (getChar(stream) == c) {
        getNextChar(stream);
        return true;
    }
    return false;
}

static void push_back(stream_t &stream) {
    stream.at -= stream.at > 0;
}

// Read string
static Token lexStr(stream_t &stream, char end) {
    Token ret(TokenType::String, stream);
    char c = getChar(stream);
    while (
        c != end &&
        c != '\n' &&
        c != EOF
    ) {
        // Append
        ret.str += std::string(1, c);
        c = getNextChar(stream);
    }
    // Errors
    if (c == EOF || c == '\n') {
        err(
            ErrType::error,
            (c == EOF ?
                "unexpected EOF while parsing string (expected '"
                : "unexpected newline while parsing string (expected '\\"
            ) + std::string(1, end) + "')",
            stream
        );
        ret.errHandled = true;
        ret.tok = TokenType::Error;
    }
    // Clean up and return
    getNextChar(stream);
    return ret;
}

// Read number
static Token lexNum(stream_t &stream) {
    Token ret(TokenType::Int, stream);
    char c = getChar(stream);
    while (isdigit(c)) {
        // Add
        ret.numVal *= 10;
        ret.numVal += c - '0';
        ret.str += std::string(1, c);
        c = getNextChar(stream);
    }
    // Decimals
    if (c == '.') {
        ret.tok = TokenType::Float;
        ret.str += ".";
        c = getNextChar(stream);
        // Trailing period error
        if (!isdigit(c)) {
            err(ErrType::error, "expected a number", stream);
            err(
                ErrType::hint,
                "sack doesn't allow trailing periods, try adding a zero after",
                stream
            );
            ret.errHandled = true;
            ret.tok = TokenType::Error;
            return ret;
        }
        // Add
        ret.floatVal = ret.numVal;
        int divisor = 1;
        while (isdigit(c)) {
            ret.floatVal += (c - '0') / (float)pow(10, divisor);
            ret.str += std::string(1, c);
            divisor++;
            c = getNextChar(stream);
        }
    }
    return ret;
}

// Read identifier
static Token lexIdentifier(stream_t &stream) {
    char c = getChar(stream);
    if (!isalpha(c) && c != '_') return Token(TokenType::Error, stream);
    Token ret(TokenType::Identifier, stream);
    while (isalnum(c) || c == '_') {
        ret.str += std::string(1, c);
        c = getNextChar(stream);
    }
    return ret;
}

// Get the next token from the stream
static Token getTok(stream_t &stream) {
    char c = getChar(stream);
    getNextChar(stream);
    switch (c) {
        // Misc single chars toks, (, ), [, ], {, }, ;, ,, %
        case '(': return Token(TokenType::Lparan, stream, "(");
        case ')': return Token(TokenType::Rparan, stream, ")");
        case '[': return Token(TokenType::Lbracket, stream, "[");
        case ']': return Token(TokenType::Rbracket, stream, "]");
        case '{': return Token(TokenType::Lbrace, stream, "{");
        case '}': return Token(TokenType::Rbrace, stream, "}");
        case ';': return Token(TokenType::Semicolon, stream, ";");
        case ',': return Token(TokenType::Comma, stream, ",");
        case '%': return Token(TokenType::Modulo, stream, "%");

        // Math, mathmath, and math= ops, +, ++, +=, -, --, -=, *, *=, /, /=
        // +=, ++, +
        case '+': return isNext(stream, '=') ?
            Token(TokenType::Plus_equals, stream, "+=")
                : isNext(stream, '+') ?
                Token(TokenType::Plus_plus, stream, "++") :
                Token(TokenType::Plus, stream, "+");
        // -=, --, -
        case '-': return isNext(stream, '=') ?
            Token(TokenType::Minus_equals, stream, "-=")
                : isNext(stream, '-') ?
                Token(TokenType::Minus_minus, stream, "--") :
                Token(TokenType::Minus, stream, "-");
        // *=, *
        case '*': return isNext(stream, '=') ?
            Token(TokenType::Times_equals, stream, "*=") :
            Token(TokenType::Times, stream, "*");
        // *=, *
        case '/': return isNext(stream, '=') ?
            Token(TokenType::Div_equals, stream, "/=") :
            Token(TokenType::Div, stream, "/");

        // Comparison ops, =, ==, !, !=, >, >=, <, <=
        // ==, =
        case '=': return isNext(stream, '=') ?
            Token(TokenType::Equals_equals, stream, "==") :
            Token(TokenType::Equals, stream, "=");
        // !=, !
        case '!': return isNext(stream, '=') ?
            Token(TokenType::Not_equals, stream, "!=") :
            Token(TokenType::Not, stream, "!");
        // >=, >
        case '>': return isNext(stream, '=') ?
            Token(TokenType::Gt_equals, stream, ">=") :
            Token(TokenType::Gt, stream, ">");
        // <=, <
        case '<': return isNext(stream, '=') ?
            Token(TokenType::Lt_equals, stream, "<=") :
            Token(TokenType::Lt, stream, "<");

        // Logical ops, &&, ||, ^^
        // &&
        case '&': return isNext(stream, '&') ?
            Token(TokenType::And, stream, "&&") :
            Token(TokenType::Error, stream, "<lexing error>");
        // ||
        case '|': return isNext(stream, '|') ?
            Token(TokenType::Or, stream, "||") :
            Token(TokenType::Error, stream, "<lexing error>");
        // ^^
        case '^': return isNext(stream, '^') ?
            Token(TokenType::Xor, stream, "^^") :
            Token(TokenType::Error, stream, "<lexing error>");

        // Comments
        case '#':
            while (!isNext(stream, '\n') && getChar(stream) != EOF) {
                getNextChar(stream);
            }
            return getTok(stream);
        // Strings
        case '"':
            return lexStr(stream, '"');
        case '\'':
            return lexStr(stream, '\'');
        // Numbers
        case '0' ... '9':
            push_back(stream);
            return lexNum(stream);

        // Whitespace
        case '\n':
        case ' ':
        case '\r':
        case '\t':
            return getTok(stream);

        // EOF
        case EOF:
            return Token(TokenType::Eof, stream);

        // Identifiers and invalid tokens
        default:
            push_back(stream);
            Token ret = lexIdentifier(stream);
            #ifdef FUNCTI
            if (ret.str == "functi") {
            #else
            if (ret.str == "func") {
            #endif
                ret.tok = TokenType::Func;
            } else if (ret.str == "let") {
                ret.tok = TokenType::Let;
            } else if (ret.str == "return") {
                ret.tok = TokenType::Return;
            } else if (ret.str == "in") {
                ret.tok = TokenType::In;
            } else if (ret.str == "if") {
                ret.tok = TokenType::If;
            } else if (ret.str == "else") {
                ret.tok = TokenType::Else;
            } else if (ret.str == "loop") {
                ret.tok = TokenType::Loop;
            } else if (ret.str == "import") {
                ret.tok = TokenType::Import;
            } else if (ret.str == "none") {
                ret.tok = TokenType::None;
            } else if (ret.str == "true" || ret.str == "false") {
                ret.tok = TokenType::Bool;
            }
            return ret;
    }
}

// Tokenize an entire file
std::vector<Token> tokenize(stream_t &stream) {
    std::vector<Token> ret = {};
    while (getChar(stream) != EOF) {
        Token token = getTok(stream);
        // Syntax errors
        if (token.tok == TokenType::Error) {
            if (token.errHandled == false)
                err(ErrType::error, "illegal symbol while parsing", stream);
            return {};
        }
        ret.push_back(token);
    }
    ret.push_back(Token(TokenType::Eof, stream));
    return ret;
}

File: src/lexer.rs
use logos::Logos;

use crate::common::{err, ErrType, Stream};

#[derive(Logos, Debug, PartialEq, Clone)]
#[logos(subpattern numbers = r"[0-9]((_?[0-9])*)?")]
// The token enum
pub enum TokenType {
    // Special tokens
    Invalid,
    Eof,
    #[regex(r"(#[^\n]*)|([ \t\f\r]+)")]
    Skipped,
    #[token("\n")]
    Newline,
    // Literals/data types
    #[regex(r"[a-zA-Z_][a-zA-Z0-9_]*", |lex| lex.slice().to_string())]
    Identifier(String),
    #[regex(
        "'[^']*'", |lex| lex.slice()[1..lex.slice().len()-1].to_string())
    ]
    #[regex(
        "\"[^\"]*\"", |lex| lex.slice()[1..lex.slice().len()-1].to_string()
    )]
    Str(String),
    #[regex(r"(?&numbers)", |lex| lex.slice().parse().ok())]
    Int(i32),
    #[regex(r"(?&numbers)\.(?&numbers)", |lex| lex.slice().parse().ok())]
    Float(f32),
    #[regex("(true)|(false)", |lex| lex.slice() == "true")]
    Bool(bool),
    #[token("none")]
    None,
    // Valid byte
    #[regex(
        "0b[01]([01]?){7}",
        |lex| u8::from_str_radix(&lex.slice()[2..], 2).ok()
    )]
    // Invalid byte
    #[regex(
        "0b[01]{8}[01]+", |_| Err(())
    )]
    Byte(u8),
    // Misc
    #[token("(")]
    Lparan,
    #[token(")")]
    Rparan,
    #[token("[")]
    Lbracket,
    #[token("]")]
    Rbracket,
    #[token("{")]
    Lbrace,
    #[token("}")]
    Rbrace,
    #[token(";")]
    Semicolon,
    #[token(":")]
    Colon,
    #[token(",")]
    Comma,
    #[token("=")]
    Equals,
    #[token("++")]
    PlusPlus,
    #[token("--")]
    MinusMinus,
    // Keywords
    #[regex("func(ti)?", |lex| lex.slice() == "functi")]
    Func(bool),
    #[token("let")]
    Let,
    #[token("return")]
    Return,
    #[token("in")]
    In,
    #[token("if")]
    If,
    #[token("else")]
    Else,
    #[token("loop")]
    Loop,
    #[token("while")]
    While,
    #[token("continue")]
    Continue,
    #[token("break")]
    Break,
    #[token("import")]
    Import,
    // Basic operators
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Times,
    #[token("/")]
    Div,
    #[token("%")]
    Modulo,
    // Basic Operators =
    #[token("+=")]
    PlusEquals,
    #[token("-=")]
    MinusEquals,
    #[token("*=")]
    TimesEquals,
    #[token("/=")]
    DivEquals,
    #[token("%=")]
    ModEquals,
    // Logical operators
    #[token("&&")]
    And,
    #[token("||")]
    Or,
    #[token("^^")]
    Xor,
    #[token("!")]
    Not,
    // Comparison operators
    #[token("==")]
    EqualsEquals,
    #[token("!=")]
    NotEquals,
    #[token("<")]
    Lt,
    #[token(">")]
    Gt,
    #[token("<=")]
    LtEquals,
    #[token(">=")]
    GtEquals
}

// Token, has a token plus everything needed for errors
#[derive(Clone)]
pub struct Token {
    pub stream: Stream,
    pub token: TokenType,
    pub str: String,
}

impl std::fmt::Debug for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{:?}", self.token)
    }
}

pub fn lex(
    src: &str, name: String, print_err: bool, color: bool
) -> Option<Vec<Token>> {
    let mut lex = TokenType::lexer(src);
    let mut ret: Vec<Token> = vec![];
    // Lines
    let lines = src.lines().collect::<Vec<&str>>();
    if lines.is_empty() {
        return Some(vec![]);
    }
    // Stream (for errors)
    let mut stream = Stream{
        name, rat: 0, at: 0, line: 1, size: 0,
    };
    let mut lastat = 0;
    let mut tok = lex.next();
    while tok.is_some() {
        stream.size = lex.span().end - lex.span().start;
        stream.rat = lex.span().start;
        stream.at = lex.span().start - lastat;
        if tok.clone().unwrap().is_err() {
            if !print_err {
                // The REPL lexes for highlighting, not syntax
                ret.push(Token{
                    token: TokenType::Invalid,
                    stream: stream.clone(),
                    str: lex.slice().to_string()
                });
                tok = lex.next();
                continue;
            }
            err(&stream, "failure to lex", ErrType::Err, color);
            return None;
        } else {
            let token = tok.unwrap().unwrap();
            if token == TokenType::Newline {
                // Bump line
                stream.line += 1;
                lastat = lex.span().start + 1;
            }
            if let TokenType::Str(ref str) = token {
                // Lines can have multiple lines in them
                let newlines = str.matches("\n").count();
                stream.line += newlines;
                lastat = lex.span().start + newlines;
            }
            if let TokenType::Newline | TokenType::Skipped = token {
                if print_err {
                    tok = lex.next();
                    continue;
                }
            }
            ret.push(Token{
                token, stream: stream.clone(),
                str: lex.slice().to_string()
            });
        }
        tok = lex.next();
    }
    ret.push(Token{
        token: TokenType::Eof,
        stream,
        str: "".to_string(),
    });
    return Some(ret);
}
